### ISSUE TITLE
AutoN-BEATS/AutoN-HITS Cross Validation Function

### ISSUE THREAD

**Issue by sdmishra123** [Nixtla/neuralforecast#568]
Hello Team,

I have a dataset that I have split into the train, valid, and test set 60-20-20%, respectively.
This is how I define my model search space and the model:-

nbeatsx_config = {    
      "learning_rate"      : tune.loguniform(1e-5, 1e-1),                                                                       # Initial Learning rate
       "max_steps"          : tune.choice([1000]),                                                                                 # Number of SGD step
       "input_size"         : tune.choice([10*horizon]),  
      #  "scaler_type"        : tune.choice([None, "robust", "standard"]),                                                         # input_size = multiplier * horizon
      #  "batch_size"         : tune.choice([32,64,128,256]),                                                                      # Number of series in windows  
       "activation"         : tune.choice(['LeakyReLU']),                                                                          # Type of non-linear activation
      #  "n_blocks"           : tune.grid_search([ 3*[1] ,3*[5], 3*[10] ]),                                                        # Blocks per each 3 stacks
      #  "mlp_units"          : tune.grid_search( [ 3 * [[128, 128]], 3 * [[256, 256]],3 * [[512, 512]] ] ),                       # 2 512-Layers per block for each stack
       "random_seed"        : tune.choice([42]),
      #  "n_harmonics"        : tune.grid_search([2,3]),
      #  "n_polynomials"      : tune.grid_search([2,3]),
       "stack_types"        : tune.choice([ ['identity', 'identity', 'identity'] ]),
      #  "windows_batch_size" : tune.choice([128, 256, 512, 1024]),

       "hist_exog_list"   : (cannot share, company proprietary)
       "futr_exog_list"   : (cannot share, company proprietary)
}

models = [ AutoNBEATSx (h=horizon,loss=MAPE(),config=nbeatsx_config,num_samples = 10,
 search_alg=BasicVariantGenerator(random_state=42)
 
#train - 60%, valid=train=20% 
Y_df = train_fin
n_time = len(train_fin.ds.unique())
val_size = int(0.2 * n_time)
test_size = int(0.2 * n_time)
print(val_size,test_size, len(train_fin)-val_size-test_size)

nf = NeuralForecast(
    models=models,
    freq='D')

Y_hat_df = nf.cross_validation(
                              df=Y_df, 
                              val_size=val_size,
                              test_size=test_size,
                              n_windows = None                                
                              )

How does this CV rolling window forecast work for time series data? The fitted values for my best model configuration look almost perfect. The fitted values show that the model captured the time series's structural break points and slope change. So, is the model good enough to capture those structural changes? Or it's using the entire train, test, and validation set in the rolling window CV?
Can you please provide more insights on the CV and rolling window functionality of the N-BEATS and N-HITS techniques?

In the link, it says:
The cross-validation method allows you to simulate multiple historical forecasts, greatly simplifying pipelines by replacing for loops with the fit and predict methods.

With time series data, cross-validation is done by defining a sliding window across the historical data and predicting the period following it. This form of cross-validation allows us to better estimate our model’s predictive abilities across a more comprehensive range of temporal instances while keeping the data in the training set contiguous, as our models require.

The cross-validation method will use the validation set for hyperparameter selection and produce the forecasts for the test set.
https://nixtla.github.io/neuralforecast/examples/longhorizon_with_nhits.html

**Comment by cchallu** [comment]
Hi @sdmishra123! When you mention, "fitted values show that the model captured the time..." are you referring to the result data frame returned by the `cross_validation` function?

The model is only trained on the entire train data (n_time-val_size-test_size). We use the validation set to compute the performance and select the best model. The forecasts returned with the function are for the test set, which is completely unseen during training or model selection. 

The function has been thoroughly tested to ensure that there is no leakage. If you think the forecasts are too good to be true, it might be caused by some leakage from the exogenous features. Only use `futr_exog_list` for the variables for which you know the future values in the forecast horizon in advance (scheduled promotions, other forecasts, calendar features, etc.)

**Comment by sdmishra123** [comment]
I am referring to the in-sample fitted values that you get after running the below code:
fitted_values = fcst.predict_insample()
The fitted_values look perfect; how does the CV rolling window forecast work? 

Question #1 with a scenario:
Let's say I am building a model with weekly frequency, and I have 100 weeks of data, and if my split is 60-20-20%, then my train set starts from week 1 to week 60th. Then my validation set starts from week 61 to week 80 and test set from week 81 to week 100.

And when the CV function runs with n_windows=None, it will take train set from week 1 to week 60 and then make the projection for the next 1 week, compare the forecast with the actuals, and then this will become the CV fold 1 error. Then, in the next iteration, it will append the 1-week data from the validation set, exclude this week 1's data from the validation set, re-run the model, and forecast for the next week. And will it repeat this until the validation set has no data point? Then finally, use the entire train + validation set and the best config to get the test forecast.
This is my understanding; I don't know if this is right. Can you please tell me the internal working of the CV with examples for more clarity?

Question #2
So, when we set the split as 60-20-20 for the train-valid-test, it takes the first 60% of the train set (from the beginning- week 1 through week 60th), then the next 20% as a validation set, and then the remaining 20% as a test set? I assume it does so, as I am dealing with temporal data. 

The exogenous variables that we are using right now are nothing but the calendar holidays. I am passing the same thing to the historical and future exogenous parameters.

If this is a lot of information, can we schedule a call and discuss this?


I appreciate any help you can provide.

**Comment by cchallu** [comment]
It is completely normal for the fitted values on the train set to be almost perfect.

Your understandings of both questions are correct. To be precise, the re-run is only the forward (prediction) step; models are not retrained during validation. Also, I am unsure what you mean by "exclude this week 1's data from the validation set". To make each prediction, the model will use up to the latest data available (as you correctly mentioned with the append 1-week data); no data is dropped from the beginning of the train set.

You can join our Slack community (link in readme) and message me directly.
